source/common/ssl/context_manager_impl.cc

```diff
@@ -20,23 +21,93 @@ void ContextManagerImpl::releaseContext(Context* context) {
   contexts_.remove(context);
 }
 
+void ContextManagerImpl::releaseServerContext(ServerContext* context,
+                                              const std::string& listener_name,
+                                              const std::vector<std::string>& server_names) {
+  std::unique_lock<std::mutex> lock(contexts_lock_);
+
+  // Remove mappings.
+  if (server_names.empty()) {
+    if (map_exact_[listener_name][EMPTY_STRING] == context) {
+      map_exact_[listener_name][EMPTY_STRING] = nullptr;
+    }
+  } else {
+    for (const auto& name : server_names) {
+      if (name.size() > 2 && name[0] == '*' && name[1] == '.') {
+        if (map_wildcard_[listener_name][name] == context) {
+          map_wildcard_[listener_name][name] = nullptr;
+        }
+      } else {
+        if (map_exact_[listener_name][name] == context) {
+          map_exact_[listener_name][name] = nullptr;
+        }
+      }
+    }
+  }
+
+  // context may not be found, in the case that a subclass of Context throws
+  // in it's constructor.  In that case the context did not get added, but
+  // the destructor of Context will run and call releaseContext().
+  contexts_.remove(context);
+}
+
 ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
                                                             ClientContextConfig& config) {
-
   ClientContextPtr context(new ClientContextImpl(*this, scope, config));
   std::unique_lock<std::mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
   return context;
 }
 
-ServerContextPtr ContextManagerImpl::createSslServerContext(Stats::Scope& scope,
-                                                            ServerContextConfig& config) {
-  ServerContextPtr context(new ServerContextImpl(*this, scope, config, runtime_));
+ServerContextPtr
+ContextManagerImpl::createSslServerContext(const std::string& listener_name,
+                                           const std::vector<std::string>& server_names,
+                                           Stats::Scope& scope, ServerContextConfig& config) {
+  ServerContextPtr context(
+      new ServerContextImpl(*this, listener_name, server_names, scope, config, runtime_));
   std::unique_lock<std::mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
+
+  // Save mappings.
+  if (server_names.empty()) {
+    map_exact_[listener_name][EMPTY_STRING] = context.get();
+  } else {
+    for (const auto& name : server_names) {
+      if (name.size() > 2 && name[0] == '*' && name[1] == '.') {
+        map_wildcard_[listener_name][name] = context.get();
+      } else {
+        map_exact_[listener_name][name] = context.get();
+      }
+    }
+  }
+
   return context;
 }
 
+ServerContext* ContextManagerImpl::findSslServerContext(const std::string& listener_name,
+                                                        const std::string& server_name) {
+  std::unique_lock<std::mutex> lock(contexts_lock_);
```

Comment:
It's very non-optimal that we are acquiring a lock here on every handshake, especially regardless of whether SNI is being used or not. In the interest of moving this along, I'm OK living with this for now. At minimum please put in a TODO. The quick fix is that we can make this a shared mutex and only acquire read access when we do a find. We are going to move to C++14 soon and that will become available. I would prefer that this be completely lockless within a single listener, and I can think of ways of doing this, but it would be a larger change. If you want to brainstorm on this let's chat offline or we can just leave as part of the TODO.

Reply:
for the uninitiated, why do we need locks here? is this shared global context across all threads, where the context can be updated by LDS or future SDS updates? Is that the reason for suggesting a r/w lock?

Reply:
The lock is required because we print the contexts as part of admin output, and contexts come and go with listeners. C++14 support has merged so this should be changed to a shared_mutex (I think C++14 only has the timed one, but it works in not timed mode). We can fix the need for a lock in this flow at all later and leave a TODO.

Reply:
Done.